### ISSUE TITLE
blade: Text is rendered either too thick or too thin

### ISSUE THREAD

**Issue by apricotbucket28** [zed-industries/zed#13564]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Ever since https://github.com/zed-industries/zed/commit/a89dc8c42efff8d746088a7b79fae14bf0d4eb78, text rendering has either been to thick or too thin, depending on the theme.

## Screenshots
_(they're not mine, I just took them from the [Discord channel](https://discord.com/channels/869392257814519848/1204679850208657418/1238060127328927856))_

### Dark theme
Before
![image](https://github.com/zed-industries/zed/assets/71973804/4e679e91-c073-4864-908a-973b97d89cfe)
After
![image](https://github.com/zed-industries/zed/assets/71973804/10c86729-a0d1-4664-9da0-74a1758c0edb)

### Light theme
Before
![image](https://github.com/zed-industries/zed/assets/71973804/cf2cd9b8-5f10-434c-bfb5-8d6c1f854512)
After
![image](https://github.com/zed-industries/zed/assets/71973804/cde52e4b-4d2b-448c-80e8-59502070b560)

### Environment

Zed: v0.142.0 (Zed Nightly eb914682b39c154dca8d6352184afbf89b00f782)
OS: Linux Wayland fedora 40
Memory: 30.5 GiB
Architecture: x86_64

Since this is an issue with the Blade renderer, I can also reproduce on Windows.

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by ajstarks** [comment]
Workaround:
setting ```"buffer_font_weight": 600``` works for light mode themes.

**Comment by ciaran-matthew-dunne** [comment]
Following on from #14169, it appears Zed is having issues with font weights. Changing the buffer font weight is ineffective until 600, which also looks badly rendered.

![Screenshot from 2024-09-09 16-50-50](https://github.com/user-attachments/assets/c774ed4c-02d4-4cd0-a5e2-7409510b6d90)
![Screenshot from 2024-09-09 16-55-32](https://github.com/user-attachments/assets/3e43e5d7-082d-437f-8bb8-677359f874e0)

[Zed.log](https://github.com/user-attachments/files/16932425/Zed.log)

**Comment by landersson** [comment]
Agreed. Font's are way to skinny by default on Linux... With "ui_font_weight": 600, the font is instead too thick and generally looks a bit off.

**Comment by donjan** [comment]
Still the case with Zed 0.163.3. I can't make it look as good as other editors (or almost any other application) on Linux.
Image shows VSCode | Zed | Geany:
![fonts](https://github.com/user-attachments/assets/115e1c92-1c6a-4719-8521-84e3e8f55488)
Reportedly this might improve if we move to a different text system, as suggested in https://github.com/zed-industries/zed/issues/13951

**Comment by lnicola** [comment]
I think this is a relatively well-known issue related to linear ("correct") blending of text. I couldn't find a better source, but https://freetype.org/freetype2/docs/hinting/text-rendering-general.html touches on this.

**Comment by chris-morgan** [comment]
Some interesting related art: https://ghostty.org/docs/install/release-notes/1-1-0#macos:-alpha-blending-improvements. They’ve just switched from sRGB blending to P3 blending on macOS which fixes the fringing problems and matches typical platform behaviour there, but have also added two other blending modes: `linear`, the “correct” thing which all of us here agree is bad (that’s why we’re in this issue!); and `linear-corrected`, which I suspect would be my favourite of them. At present this gamma correction work is macOS-only; it’s implemented there in a Metal shader, with https://github.com/ghostty-org/ghostty/pull/5401/files#diff-873937e2e228adeceb577fbf459d00cf50f5bbdfcb23ea09cf93168b71af2a16 containing the `linear-corrected` mode, including this comment:

> Short explanation of how this works:
>
> We get the luminances of the foreground and background colors, and then unlinearize them and perform blending on them. This gives us our desired luminance, which we derive our new alpha value from by mapping the range [bg_l, fg_l] to [0, 1], since our final blend will be a linear interpolation from bg to fg.
>
> This yields virtually identical results for grayscale blending, and very similar but non-identical results for color blending.

**Comment by everdrone** [comment]
Having the same issue where I cannot match the same font weight I use in VSCode, SF Mono looks identical with a weight of 400 or 500. 600 is too thick, 400 is too thin, 500 just "doesn't work"

**Comment by javea7171** [comment]
Same issue, there is no visual difference in font weight between 400 and 500

**Comment by manaskarekar** [comment]
Another confirmation, observed on two machines.

**Comment by neyb** [comment]
there seems to be a threashold at 501 

https://github.com/user-attachments/assets/f166bde8-0e7a-4109-96e9-d56e78084721

**Comment by neyb** [comment]
So 1 year, later text rendering is still broken... Annoying for a text focused ide...

I guess linux support limits on shipping broken binaries...

Anyway... I'll try to look at this...

**Comment by landersson** [comment]
Proper text rendering under Linux is probably the main reason why I haven't seriously looked into zed as an alternative to vim/neovim.

**Comment by kocsis1david** [comment]
> I think this is a relatively well-known issue related to linear ("correct") blending of text. I couldn't find a better source, but https://freetype.org/freetype2/docs/hinting/text-rendering-general.html touches on this.

As I understand this, Zed currently does the correct thing by doing alpha blending with linear colors, but it doesn't do the stem darkening workaround.

**Comment by neyb** [comment]
> Proper text rendering under Linux is probably the main reason why I haven't seriously looked into zed as an alternative to vim/neovim.

I understand. I mean I set it to 900 and I'm dealing with it... But it's a real problem.

More over, this issue is questioning the linux support quality... I mean it's kind of a text editor... Having text badly rendered should be a top priority.

**Comment by neyb** [comment]
It seems some font are more subject than others : using "Fira Code" seems to have a lot more "steps".

**Comment by neyb** [comment]
Curiously the reporter of #14175 seems to be using macos. Might it be more a font problem of zed Plex mono?

**Comment by lpillmann** [comment]
> Might it be more a font problem of zed Plex mono?

Yes, I also think it is related to the font. I installed JetBrains Mono and it looks good and the weights work as expected

400:

<img width="539" height="342" alt="Image" src="https://github.com/user-attachments/assets/6457db40-2322-4b45-a074-17d78554c443" />

500:

<img width="568" height="325" alt="Image" src="https://github.com/user-attachments/assets/7a254143-0ca5-4619-a67c-b8db489d5e50" />

600:

<img width="581" height="333" alt="Image" src="https://github.com/user-attachments/assets/068ac38a-025c-4989-a656-2aab46847b41" />

**Comment by SomeoneToIgnore** [comment]
Curious if https://github.com/zed-industries/zed/pull/38967 helps anyone enough, or is there more, font-related issues, we should look for?

**Comment by lnicola** [comment]
Yeah, it's a lot better. I think I'd still like more contrast, but that's probably the theme (middle-click these to open in a new tab):

<img width="721" height="296" alt="Image" src="https://github.com/user-attachments/assets/e001e64f-5467-4bd0-af0c-667d460d4cc0" />

<img width="636" height="439" alt="Image" src="https://github.com/user-attachments/assets/99a0ebb1-d393-449e-b1e0-4e4433359d55" />

<img width="615" height="364" alt="Image" src="https://github.com/user-attachments/assets/fd082b43-6de0-49e2-a4f0-148b0e863f13" />

**Comment by kocsis1david** [comment]
With this change the text in the light mode now looks thicker than in the dark mode.

**Comment by reflectronic** [comment]
That might just be the theme. With pure black-and-white, they look balanced to me (if anything, the dark-on-light is maybe a hair thinner than the light-on-dark text).

(Open this picture at a 1:1 scale to see what I mean.) 
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/5bccbdd9-9a5a-4531-9d4a-530e8193472a" />

**Comment by kocsis1david** [comment]
When I run the `text` example, it looks good now on Linux with this change. I also rolled back before this PR, and `text` example had black text on white too thin. (But on the screenshot that was attached to the PR, I still see the dark text on light background too thick)

I'm not familiar with the code, but  `<BladeAtlas as PlatformAtlas>::get_or_insert_with` seems to write the glyph image with linear values into the `r` channel of the texture. But the `fs_mono_sprite` still did sRGB to linear conversion on already linear values before this PR, I don't understand why.

**Comment by weijarz** [comment]
My solution to this problem is to use a **bold** font. 
Recommended: https://github.com/evilmartians/mono
